becquerel/core/calibration.py

```diff
@@ -0,0 +1,787 @@
+"""Generic calibration class."""
+
+from abc import abstractmethod
+import ast
+import copy
+import asteval
+import black
+import blib2to3
+import numpy as np
+import scipy
+import scipy.optimize
+from .. import io
+
+CLIP_MAX = 1e6  # maximum value for a calibration function
+
+safe_eval = asteval.Interpreter(use_numpy=False)
+safe_eval.symtable["np"] = np
+safe_eval.symtable["numpy"] = np
+safe_eval.symtable["scipy"] = scipy
+
+
+class CalibrationError(Exception):
+    """Base class for calibration errors."""
+
+    pass
+
+
+def _check_points(points_x, points_y):
+    """Perform various checks on the sets of calibration points.
+
+    Ensure the arrays of points are both 1-D and have the same length,
+    that all values are >= 0, and then put them in the order of ascending
+    x values.
+
+    Parameters
+    ----------
+    points_x : float or array_like
+        The x-value or values of calibration points
+    points_y : float or array_like
+        The y-value or values of calibration points
+
+    Returns
+    -------
+    points_x : array_like
+        The x-value or values of calibration points
+    points_y : array_like
+        The y-value or values of calibration points
+    """
+    if points_x is None:
+        points_x = []
+    if points_y is None:
+        points_y = []
+    points_x = np.atleast_1d(points_x)
+    points_y = np.atleast_1d(points_y)
+    if points_x.ndim != 1:
+        raise CalibrationError(f"Calibration x points must be 1-D: {points_x}")
+    if points_y.ndim != 1:
+        raise CalibrationError(f"Calibration y points must be 1-D: {points_y}")
+    if len(points_x) != len(points_y):
+        raise CalibrationError(
+            f"Number of x and y calibration points must match: {len(points_x)}, {len(points_y)}"
+        )
+    # sort points in increasing order of x
+    i = np.argsort(points_x)
+    points_x = points_x[i]
+    points_y = points_y[i]
+    # check all values are positive
+    if not np.all(points_x >= 0):
+        raise CalibrationError(f"All calibration x points must be >= 0: {points_x}")
+    if not np.all(points_y >= 0):
+        raise CalibrationError(f"All calibration y points must be >= 0: {points_y}")
+    return points_x, points_y
+
+
+class Calibration(object):
+    """Base class for calibrations.
+
+    A calibration is a nonnegative scalar function of a nonnegative scalar
```

Comment:
This is not quite true if e.g. a linear calibration has a negative offset.

Reply:
Yes, I noticed this later when coding -- I had to remove a non-negativity check on the function output.

Reply:
Just fixed this docstring.